File: src-tauri/src/application/use_cases/scan_sprays.rs
//! Use case: scan a spray library and produce thumbnails.

use std::sync::Arc;

use crate::domain::entities::Spray;
use crate::domain::error::AppResult;
use crate::domain::repositories::SprayRepository;

pub struct ScanSprays {
    sprays: Arc<dyn SprayRepository>,
}

impl ScanSprays {
    pub fn new(sprays: Arc<dyn SprayRepository>) -> Self {
        Self { sprays }
    }

    pub fn scan(&self, library_dir: &str) -> AppResult<Vec<Spray>> {
        let mut sprays = self.sprays.scan(library_dir)?;
        // Deterministic, case-insensitive ordering for a stable grid.
        sprays.sort_by_key(|s| s.name.to_lowercase());
        Ok(sprays)
    }

    pub fn thumbnail(&self, vtf_path: &str) -> AppResult<String> {
        self.sprays.thumbnail(vtf_path)
    }
}

File: src/application/use-cases/scan-sprays.ts
import type { Spray } from "@/domain/entities/spray";
import type { SprayRepository } from "@/domain/repositories/spray-repository";

/** Scan a library and filter the results by a search query. */
export class ScanSprays {
  constructor(private readonly repo: SprayRepository) {}

  async execute(libraryDir: string): Promise<Spray[]> {
    return this.repo.scan(libraryDir);
  }

  thumbnail(vtfPath: string): Promise<string> {
    return this.repo.thumbnail(vtfPath);
  }

  delete(vtfPath: string, vmtPath: string | null): Promise<void> {
    return this.repo.delete(vtfPath, vmtPath);
  }

  /** Pure helper: instant client-side filtering by name. */
  static filter(sprays: Spray[], query: string): Spray[] {
    const q = query.trim().toLowerCase();
    if (!q) return sprays;
    return sprays.filter((s) => s.name.toLowerCase().includes(q));
  }
}
